httpx: add page (#12279)
tldr-pages/tldr: pages/common/httpx.md

# httpx

> A fast and multi-purpose HTTP toolkit written in Go to run multiple probes at once.

> > (Note that this is not the Python HTTPX which has the exact same command name).

Author: ```suggestion
> Note: This page uses the same command name as the Python HTTPX library, but it's not related.
```

Reviewer: ```suggestion
> Note: not to be confused with Python HTTPX which has the exact same command name.
```
How about this?

> > (Note that this is not the Python HTTPX which has the exact same command name).

Author: I suggested an alternate suggestion below which I think is provides more clarity.

> More information: <https://github.com/projectdiscovery/httpx>.

- Run a probe against a URL, host, IP Address or subnet (CIDR notation) showing probe status:

`httpx -probe -u {{url|host|ipaddress|subnet_with_cidr}}`

- Run a probe against multiple hosts showing [s]tatus [c]ode with input from `subfinder`:

`subfinder -d {{example.com}} | httpx -sc`

- Run a [r]ate [l]imited probe against a [l]ist of hosts from a file showing [t]echnology [d]etected and [r]esponse [t]ime:

`httpx -rl {{150}} -l {{file_containing_newline_separate_hosts}} -td -rt`

- Run a probe against a URL showing webpage title, CDN/WAF in use, and page content hash:

`httpx -u {{url}} -title -cdn -hash {{sha256}}`

- Run a probe against a list of hosts with custom defined [p]orts and timeout after certain seconds:

> `httpx -probe -u {{host1}},{{host2}},{{hostN}} -p http:{{80,8000-8080}},https:{{443,8443}} -timeout {{10}}`

Author: Wrong syntax if used in `httpx` as hosts must be separated by commas. Here is the output from `httpx`

```console
# WITH SPACE SEPARATED, ONLY THE FIRST HOST IS SCANNED

$ httpx -u www.google.com www.yahoo.com www.github.com -p http:80,81,https:443,8443 -timeout 5

    __    __  __       _  __
   / /_  / /_/ /_____ | |/ /
  / __ \/ __/ __/ __ \|   /
 / / / / /_/ /_/ /_/ /   |
/_/ /_/\__/\__/ .___/_/|_|
             /_/

                projectdiscovery.io

[INF] Current httpx version v1.4.0 (latest)
https://www.google.com
$

###############################################################

# USING THE CORRECT SYNTAX ALL HOSTS ARE SCANNED

$ httpx -u www.google.com,www.yahoo.com,www.github.com -p http:80,81,https:443,8443 -timeout 5 

    __    __  __       _  __
   / /_  / /_/ /_____ | |/ /
  / __ \/ __/ __/ __ \|   /
 / / / / /_/ /_/ /_/ /   |
/_/ /_/\__/\__/ .___/_/|_|
             /_/

                projectdiscovery.io

[INF] Current httpx version v1.4.0 (latest)
http://www.github.com
https://www.github.com
https://www.yahoo.com
http://www.google.com
https://www.google.com
http://www.yahoo.com
$ 
```

If we must follow the style guide here, maybe this is acceptable?

```console
`httpx -probe -u {{host1,host2,...}} -p http:{{80,8000-8080}},https:{{443,8443}} -timeout {{10}}`
```

> `httpx -probe -u {{host1}},{{host2}},{{hostN}} -p http:{{80,8000-8080}},https:{{443,8443}} -timeout {{10}}`

Reviewer: Oh, sorry, overlooked it, but yeah, this suggestion fits the style guide just fine.

> `httpx -probe -u {{host1}},{{host2}},{{hostN}} -p http:{{80,8000-8080}},https:{{443,8443}} -timeout {{10}}`

Author: I will push the changes following the style guide.

> `httpx -probe -u {{host1}},{{host2}},{{hostN}} -p http:{{80,8000-8080}},https:{{443,8443}} -timeout {{10}}`

Author: Fixed

- Run a probe against a list of hosts [f]iltering out [c]odes of certain responses:

> `httpx -u {{host1}},{{host2}},{{hostN}} -fc {{400,401,404}}`

Author: See above comment.

> `httpx -u {{host1}},{{host2}},{{hostN}} -fc {{400,401,404}}`

Author: Fixed

- Run a probe against a list of hosts [m]atching [c]odes of certain responses:

> `httpx -u {{host1}},{{host2}},{{hostN}} -mc {{200,301,304}}`

Author: See above comment.

> `httpx -u {{host1}},{{host2}},{{hostN}} -mc {{200,301,304}}`

Author: Fixed

> - Run a probe against a URL [s]aving [s]creenshots of certain paths, with [s]creenshot [t]imeouts (assets are saved in ./output):

Reviewer: ```suggestion
- Run a probe against a URL [s]aving [s]creenshots of certain paths, with [s]creenshot [t]imeouts (assets are saved in `./output`):
```
From the [style guide](https://github.com/tldr-pages/tldr/blob/main/contributing-guides/style-guide.md#special-cases):
> Technical wording on description lines should use the `backtick` syntax.
> Use backticks on the following:
> 
> - Paths, e.g. `package.json`, `/etc/package.json`.
> - Extensions, e.g. `.dll`.
> - Commands, e.g. `ls`.
> - Standard streams: `stdout`, `stdin`, `stderr`. **Do not** use the full names (e.g. standard output).
